llvm/lib/IR/Constants.cpp

```diff
@@ -2067,6 +2067,10 @@ Constant *ConstantExpr::getBitCast(Constant *C, Type *DstTy,
 
 Constant *ConstantExpr::getAddrSpaceCast(Constant *C, Type *DstTy,
                                          bool OnlyIfReduced) {
+  // Skip cast if types are identical
```

Comment:
Don't think we want to change this function. Was this giving incorrect results without this? Figured a no-op would be valid.

Reply:
`CastInst::castIsValid` returns false for addrspace casts if two pointers share the same addrspace.

https://github.com/llvm/llvm-project/blob/2de269a641e4ffbb7a44e559c4c0a91bb66df823/llvm/lib/IR/Instructions.cpp#L3802-L3805

Reply:
I guess leave this as it was. The addrspacecast helpers are a mess, we should clear out the vestiges of opaque pointers. The cast build helpers should accept same same address space and elide the cast, like the other cases. Any changes there should be a separate change